oracle/packages/node/tests/functions/mod.rs

```diff
@@ -0,0 +1,61 @@
+use async_trait::async_trait;
+use fuels::tx::Receipt;
+use futures::executor::block_on;
+use node::{PriceProvider, PriceUpdater};
+use std::borrow::BorrowMut;
+use std::sync::Arc;
+use std::time::Instant;
+use tokio::sync::Mutex;
+
+mod run;
+
+struct Invocation {
+    // Used to ensure the polled price is correct
+    price: u64,
+    // Used to ensure the polling delay is correct
+    time: Instant,
+}
+
+struct LoggingPriceUpdater {
+    invocations: Arc<Mutex<Vec<Invocation>>>,
+    receipts: Vec<Receipt>,
+}
+
+impl LoggingPriceUpdater {
+    pub fn new() -> Self {
+        Self {
+            invocations: Arc::new(Mutex::new(vec![])),
+            receipts: vec![],
+        }
+    }
+
+    pub fn invocations(&self) -> Arc<Mutex<Vec<Invocation>>> {
+        Arc::clone(&self.invocations)
```

Comment:
oof, so I'm aware of this area however I've not delved into it so it's a good idea to have an experienced Rust dev take a look at the implementation to see if this makes sense (not just this line but in general)

Reply:
Yeah I'll ping the rustceans in the sdk team.  @segfault-magnet wrote most of the code  :)

Reply:
It's used only in tests. It's a poor man's [spy](https://martinfowler.com/bliki/TestDouble.html):

> Spies are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.

Since the spawn oracle updater fn requires ownership over the price fetcher and price updater, you need some way of giving away the price updater but still having a way of spying on how it is being used for testing purposes. And it mustn't make the updater `!Send`. 

A couple of ways to go about it, one of those is for the spy to hold an inner collection of invocations that is behind a reference count -- i.e. has shared ownership. The Mutex part derives from the 'mustn't make the updater `!Send`' part.

Or you can use a library for generating testing doubles. Since I haven't worked with any such library in Rust, felt quicker to just DIY.